### ISSUE TITLE
训练模型时，如果后面一层的梯度为0，则前一层的梯度就变成了nan

### ISSUE THREAD

**Issue by lipan6461188** [PaddlePaddle/Paddle#29922]
PaddlePaddle版本：2.0.0-rc1
GPU: NVIDIA-SMI 440.33.01    Driver Version: 440.33.01    CUDA Version: 10.2    CUDNN: 7.4
系统：Linux ubuntu 4.4.0-131-generic #157-Ubuntu SMP Thu Jul 12 15:51:36 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
训练：单机单卡

这是我的残差层的定义：
```python
class ResBlock(paddle.nn.Layer):
    """
    Residual block
    """
    def __init__(self, dim=48, dilation=1):
        super(ResBlock, self).__init__()
        self._conv1 = paddle.nn.Conv2D(in_channels=dim, out_channels=dim, kernel_size=3, padding=1+(dilation-1), dilation=dilation)
        self._conv2 = paddle.nn.Conv2D(in_channels=dim, out_channels=dim, kernel_size=5, padding=2+(dilation-1), dilation=dilation)
        self._in1 = paddle.nn.InstanceNorm2D(num_features=dim)
        self._in2 = paddle.nn.InstanceNorm2D(num_features=dim)
        self._dropout = paddle.nn.Dropout2D(p=0.25, data_format='NCHW')
        self._elu1 = paddle.nn.ELU()
        self._elu2 = paddle.nn.ELU()
        
    def forward(self, x):
        y = self._elu1(x)
        y = self._in1(y)
        y = paddle.triu(y, diagonal=2) #y.triu(diagonal=2)
        y = self._conv1(y)
        y = self._elu2(y)
        y = self._in2(y)
        y = paddle.triu(y, diagonal=2) #y.triu(diagonal=2)
        y = self._dropout(y)
        y = self._conv2(y)
        assert x.shape == y.shape
        return paddle.add(x, y)
```

因为这个层的输入和输出的维度是一样的，我在实际使用中把多个残差层串在了一起：
```python
seq_modules_list = [ InputConvUnit(out_dim=chanNum) ]
for i in range(resBlockNum):
      seq_modules_list.append( ResBlock(dim=chanNum, dilation=dilaFactor) )
```
**问题**：在训练的过程中，我发现只要后面的一层梯度变成了0，前面一层的梯度就会变成nan。下面的代码输出后面一层的参数梯度：
```python
params = getattr(model, '3').parameters() # 层3是后面一层
for i,para in enumerate(params):
    zero_num = (para.grad==0).astype('float32').sum()
    nan_num = np.isnan(para.grad).astype('float32').sum()
    print(f"Param: {i}, Zero_Num: {zero_num}, Nan_Num: {nan_num}")
```
输出为：
```
Param: 0, Zero_Num: 5184.0, Nan_Num: 0.0
Param: 1, Zero_Num: 12.0, Nan_Num: 0.0
Param: 2, Zero_Num: 14400.0, Nan_Num: 0.0
Param: 3, Zero_Num: 0.0, Nan_Num: 0.0
Param: 4, Zero_Num: 0.0, Nan_Num: 0.0
Param: 5, Zero_Num: 0.0, Nan_Num: 0.0
Param: 6, Zero_Num: 12.0, Nan_Num: 0.0
Param: 7, Zero_Num: 12.0, Nan_Num: 0.0
```
可以看到很多参数梯度是0。下面的代码输出前面一层的参数梯度：
```python
params = getattr(model, '2').parameters() # 层2是前面一层
for i,para in enumerate(params):
    zero_num = (para.grad==0).astype('float32').sum()
    nan_num = np.isnan(para.grad).astype('float32').sum()
    print(f"Param: {i}, Zero_Num: {zero_num}, Nan_Num: {nan_num}")
```
输出为：
```
Param: 0, Zero_Num: 0.0, Nan_Num: 20736.0
Param: 1, Zero_Num: 0.0, Nan_Num: 48.0
Param: 2, Zero_Num: 12925.0, Nan_Num: 1200.0
Param: 3, Zero_Num: 0.0, Nan_Num: 1.0
Param: 4, Zero_Num: 0.0, Nan_Num: 48.0
Param: 5, Zero_Num: 0.0, Nan_Num: 0.0
Param: 6, Zero_Num: 0.0, Nan_Num: 48.0
Param: 7, Zero_Num: 0.0, Nan_Num: 48.0
```
可以看到很多层的参数梯度变成了nan。此时如果运行`optimizer.step()`，则参数也会变成nan。

当后一层的梯度为0的时候，前一层的梯度更新为nan是否是一个bug?

**Comment by lipan6461188** [comment]
我仔细研究发现上面的问题是不成立的，我再训练的时候之所以会出现nan是因为我的模型顺序是 `FC->Activation->Dropout->BatchNorm`，我改成`FC->BatchNorm->Activation->Dropout`以后就没有问题了。
